src/model.jl

```diff
@@ -1097,6 +1098,43 @@ This is the `Model` with the lowest model number.
 """
 defaultmodel(struc::MolecularStructure) = first(sort(collect(values(models(struc)))))
 
+"""
+    specialize_resnames!(r)
+
+If `r` is a histidine residue, rename it to reflect a specific protonation configuration (HID,
+HIE, or HIP) based on the presence of HD1 and HE2 atoms. For any other residue,
+`r` will be unchanged.
+
+Hydrogens must have been assigned prior to calling this function.
+"""
+function specialize_resnames!(r::Residue)
+    rname = resname(r)
+    if rname == "HIS"
+        hd1 = findatombyname(r, "HD1"; strict=false)
+        he2 = findatombyname(r, "HE2"; strict=false)
+        if !isnothing(hd1) && !isnothing(he2)
+            r.name = "HIP"
+        elseif !isnothing(hd1)
+            r.name = "HID"
+        elseif !isnothing(he2)
+            r.name = "HIE"
+        else
+            error("at least one of HD1 or HE2 must be present")
+        end
+    end
+    # N and C termini
+    if haskey(r.atoms, "H3") && !haskey(residuedata[rname].atoms, "H3") && haskey(residuedata, "N" * rname)
+        r.name = "N" * rname
+    elseif haskey(r.atoms, "OXT") && !haskey(residuedata[rname].atoms, "OXT") && haskey(residuedata, "C" * rname)
+        r.name = "C" * rname
+    end
+    return r
+end
+specialize_resnames!(r::DisorderedResidue) = (foreach(specialize_resnames!, collectresidues(r; expand_disordered=true)); r)
```

Comment:
I think for `DisorderedResidue` you should also update the key in the `names` dictionary, otherwise it will go out of sync with the residue name.